File: examples/hello_world_labels.py
from main import *


def main_run() -> None:
    # 1. Create a printer device with a custom callback that prints to the terminal
    printer = PrinterDevice(output_callback=lambda char: print(char, end=""))

    # 2. Initialize the machine with the printer device attached to port 0x02
    machine = Machine.create(
        address_lines=16,
        data_lines=8,
        devices=[(printer, [0x02])]
    )

    # 3. Create the 8085 assembly program using label referencing
    program = Program([
        Instruction(Opcode.LXI, "STR_DATA"),          # HL points to start of string (using LXI label)

        # LOOP: Load char and check for null
        Instruction(Opcode.MOV_A_M, label="LOOP"),
        Instruction(Opcode.CPI, Data.byte(0x00)),
        Instruction(Opcode.JZ, "EXIT"),               # Jump to EXIT label if character is null

        # Output character, advance pointer, and repeat
        Instruction(Opcode.OUT, Data.byte(0x02)),
        Instruction(Opcode.INX_HL),                   # Point HL to next memory location
        Instruction(Opcode.JMP, "LOOP"),              # Loop back to LOOP label

        # EXIT: Halt execution
        Instruction(Opcode.HLT, label="EXIT"),

        # STR_DATA: Placeholder instruction representing the start of string data
        Instruction(Opcode.NOP, label="STR_DATA")
    ])

    # 4. Load program at entry address 0x0000
    machine.load(program, Mem(0x0000))

    # 5. Load the string message into RAM at the resolved address of "STR_DATA"
    # Instruction sizes: LXI (3), MOV_A_M (1), CPI (2), JZ (3), OUT (2), INX_HL (1), JMP (3), HLT (1).
    # Total size of executable code before STR_DATA is 16 bytes.
    # So STR_DATA is located at 0x0010.
    str_addr = Mem(0x0010)
    message = b"Hello, World using Labels!\n"
    for i, byte_val in enumerate(message):
        machine.ram.write(Mem(str_addr + i), Data.byte(byte_val))
    # Null terminator
    machine.ram.write(Mem(str_addr + len(message)), Data.byte(0x00))

    # 6. Run the machine
    print("--- Executing Hello World with Labels ---")
    machine.run()
    print("-----------------------------------------")

if __name__ == "__main__":
    main_run()

File: examples/hello_world_labels.rs
//! Prints "Hi Labels!" by walking a label-referenced string in RAM and sending each
//! byte to a printer on I/O port 0x02. Run with `cargo run --example hello_world_labels`.

use emu8085::{Addr, Instruction, Machine, Opcode, Operand, PrinterDevice, Program};
use std::cell::RefCell;
use std::rc::Rc;

fn main() {
    let printed = Rc::new(RefCell::new(String::new()));
    let sink = printed.clone();
    let printer = PrinterDevice::with_callback(move |ch| {
        print!("{ch}");
        sink.borrow_mut().push(ch);
    });

    let mut machine = Machine::create(16, 8);
    machine.attach_device(Box::new(printer), &[0x02]);

    let program = Program::new(vec![
        Instruction::with(Opcode::MVI_HL, Operand::label("STR_DATA")), // LXI H, STR_DATA
        Instruction::new(Opcode::MOV_AM).labeled("LOOP"),              // A = *HL
        Instruction::with(Opcode::CPI, Operand::byte(0x00)),          // NUL?
        Instruction::with(Opcode::JZ, Operand::label("EXIT")),
        Instruction::with(Opcode::OUT, Operand::byte(0x02)),          // print A
        Instruction::new(Opcode::INX_HL),                              // HL++
        Instruction::with(Opcode::JMP, Operand::label("LOOP")),
        Instruction::new(Opcode::HLT).labeled("EXIT"),
        Instruction::new(Opcode::NOP).labeled("STR_DATA"),
    ]);

    machine.load(&program, Addr(0x0000)).expect("program compiles");

    // Place the message (NUL-terminated) at STR_DATA (0x0010).
    let message = b"Hi Labels!\n";
    for (i, &b) in message.iter().enumerate() {
        machine.ram.write(Addr(0x0010 + i as u16), b);
    }
    machine.ram.write(Addr(0x0010 + message.len() as u16), 0x00);

    machine.run();
    assert_eq!(*printed.borrow(), "Hi Labels!\n");
}
